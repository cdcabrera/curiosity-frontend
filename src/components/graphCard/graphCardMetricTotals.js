import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardBody, CardFooter, CardHeader, CardTitle, Title } from '@patternfly/react-core';
import _camelCase from 'lodash/camelCase';
import { useProductGraphTallyQuery } from '../productView/productViewContext';
import { useGraphCardContext, useMetricsSelector } from './graphCardContext';
import { Loader, SkeletonSize } from '../loader/loader';
import { toolbarFieldOptions } from '../toolbar/toolbarFieldRangedMonthly';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';

/**
 * @memberof GraphCard
 * @module GraphCardMetricTotals
 */

/**
 * Display totals for a single metric.
 *
 * @param {object} props
 * @param {React.ReactNode} props.children
 * @param {Function} props.useGraphCardContext
 * @param {Function} props.useMetricsSelector
 * @param {Function} props.useProductGraphTallyQuery
 * @returns {React.ReactNode}
 */
const GraphCardMetricTotals = ({
  children,
  useGraphCardContext: useAliasGraphCardContext,
  useMetricsSelector: useAliasMetricsSelector,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery
}) => {
  const { settings = {} } = useAliasGraphCardContext();
  const { [RHSM_API_QUERY_SET_TYPES.START_DATE]: startDate } = useAliasProductGraphTallyQuery();
  const { pending, error, fulfilled, dataSets = [] } = useAliasMetricsSelector();
  const { data = [], id: chartId, metric: metricId, meta = {} } = dataSets[0] || {};
  const { date: lastDate, hasData: lastHasData, y: lastValue } = data[data.length - 1] || {};

  const {
    date: currentDate,
    hasData: currentHasData,
    y: currentValue
  } = data.find(({ isCurrentDate }) => isCurrentDate === true) || {};

  const { totalMonthlyDate: monthlyDate, totalMonthlyHasData: monthlyHasData, totalMonthlyValue: monthlyValue } = meta;

  const { title: selectedMonth, isCurrent } =
    toolbarFieldOptions.find(
      option => option.title === startDate || option.value.startDate.toISOString() === startDate
    ) || {};

  const dailyDate = isCurrent ? currentDate : lastDate;
  const dailyHasData = isCurrent ? currentHasData : lastHasData;
  const dailyValue = isCurrent ? currentValue : lastValue;

  if (settings?.isMetricDisplay && settings?.cards?.length) {
    return (
      <div data-test={`graphMetricTotals-${_camelCase(metricId)}`} className="curiosity-usage-graph__totals">
        <div>
          <div className="curiosity-usage-graph__totals-column">
            {settings?.cards?.map(({ header, body, footer }, index) => (
              <Card
                isPlain
                data-test={`graphMetricTotalsCard-${index}`}
                className={`curiosity-usage-graph__totals-column-card ${(error && 'blur') || ''}`}
              >
                <CardHeader>
                  <CardTitle>
                    <Title headingLevel="h2" size="md">
                      {pending && <Loader variant="skeleton" skeletonProps={{ size: SkeletonSize.lg }} />}
                      {fulfilled &&
                        ((typeof header === 'function' &&
                          header({
                            chartId,
                            dailyDate,
                            dailyHasData,
                            dailyValue,
                            metricId,
                            monthlyDate,
                            monthlyHasData,
                            monthlyValue,
                            selectedValue: selectedMonth
                          })) ||
                          header)}
                    </Title>
                  </CardTitle>
                </CardHeader>
                <CardBody>
                  <div>
                    {pending && <Loader variant="skeleton" skeletonProps={{ size: SkeletonSize.lg, height: '60px' }} />}
                    {fulfilled &&
                      ((typeof body === 'function' &&
                        body({
                          chartId,
                          dailyDate,
                          dailyHasData,
                          dailyValue,
                          metricId,
                          monthlyDate,
                          monthlyHasData,
                          monthlyValue,
                          selectedValue: selectedMonth
                        })) ||
                        body)}
                  </div>
                </CardBody>
                <CardFooter>
                  <div>
                    {pending && <Loader variant="skeleton" skeletonProps={{ size: SkeletonSize.lg }} />}
                    {fulfilled &&
                      ((typeof footer === 'function' &&
                        footer({
                          chartId,
                          dailyDate,
                          dailyHasData,
                          dailyValue,
                          metricId,
                          monthlyDate,
                          monthlyHasData,
                          monthlyValue,
                          selectedValue: selectedMonth
                        })) ||
                        footer)}
                  </div>
                </CardFooter>
              </Card>
            ))}
          </div>
        </div>
        <div>
          <div className="curiosity-usage-graph__totals-graph-column">{children}</div>
        </div>
      </div>
    );
  }

  return children;
};

/**
 * Prop types.
 *
 * @type {{useProductGraphTallyQuery: Function, children: React.ReactNode, useMetricsSelector: Function}}
 */
GraphCardMetricTotals.propTypes = {
  children: PropTypes.node,
  useGraphCardContext: PropTypes.func,
  useMetricsSelector: PropTypes.func,
  useProductGraphTallyQuery: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useProductGraphTallyQuery: Function, children: React.ReactNode, useMetricsSelector: Function}}
 */
GraphCardMetricTotals.defaultProps = {
  children: null,
  useGraphCardContext,
  useMetricsSelector,
  useProductGraphTallyQuery
};

export { GraphCardMetricTotals as default, GraphCardMetricTotals };
