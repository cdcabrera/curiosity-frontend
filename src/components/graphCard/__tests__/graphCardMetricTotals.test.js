import React from 'react';
import { GraphCardMetricTotals } from '../graphCardMetricTotals';

describe('GraphCardMetricTotals Component', () => {
  it('should render a basic component', async () => {
    const props = {};
    const component = await shallowHookComponent(<GraphCardMetricTotals {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should handle multiple display states', async () => {
    const props = {
      useGraphCardContext: () => ({
        settings: {
          isMetricDisplay: true,
          groupMetric: ['hello', 'world'],
          cards: [
            {
              header: 'lorem',
              body: 'ipsum',
              footer: 'dolor sit'
            }
          ]
        }
      }),
      useMetricsSelector: () => ({
        pending: true,
        error: false,
        fulfilled: false
      })
    };
    const component = await shallowHookComponent(<GraphCardMetricTotals {...props} />);

    expect(component).toMatchSnapshot('pending');

    component.setProps({
      useMetricsSelector: () => ({
        pending: false,
        error: true,
        fulfilled: false
      })
    });

    expect(component).toMatchSnapshot('error');

    component.setProps({
      useMetricsSelector: () => ({
        pending: false,
        error: false,
        fulfilled: true
      })
    });

    expect(component).toMatchSnapshot('fulfilled');
  });
});
