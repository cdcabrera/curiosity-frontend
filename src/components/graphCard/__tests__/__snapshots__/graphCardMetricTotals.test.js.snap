// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCardMetricTotals Component should handle multiple display states: error 1`] = `
<div
  className="curiosity-usage-graph__totals"
  data-test="graphMetricTotals-hello-world"
>
  <div>
    <div
      className="curiosity-usage-graph__totals-column"
    >
      <Card
        className="curiosity-usage-graph__totals-column-card blur"
        data-test="graphMetricTotalsCard-0"
        isPlain={true}
      >
        <CardHeader>
          <CardTitle>
            <Title
              headingLevel="h2"
              size="md"
            />
          </CardTitle>
        </CardHeader>
        <CardBody>
          <div />
        </CardBody>
        <CardFooter>
          <div />
        </CardFooter>
      </Card>
    </div>
  </div>
  <div>
    <div
      className="curiosity-usage-graph__totals-graph-column"
    />
  </div>
</div>
`;

exports[`GraphCardMetricTotals Component should handle multiple display states: fulfilled 1`] = `
<div
  className="curiosity-usage-graph__totals"
  data-test="graphMetricTotals-hello-world"
>
  <div>
    <div
      className="curiosity-usage-graph__totals-column"
    >
      <Card
        className="curiosity-usage-graph__totals-column-card "
        data-test="graphMetricTotalsCard-0"
        isPlain={true}
      >
        <CardHeader>
          <CardTitle>
            <Title
              headingLevel="h2"
              size="md"
            >
              lorem
            </Title>
          </CardTitle>
        </CardHeader>
        <CardBody>
          <div>
            ipsum
          </div>
        </CardBody>
        <CardFooter>
          <div>
            dolor sit
          </div>
        </CardFooter>
      </Card>
    </div>
  </div>
  <div>
    <div
      className="curiosity-usage-graph__totals-graph-column"
    />
  </div>
</div>
`;

exports[`GraphCardMetricTotals Component should handle multiple display states: pending 1`] = `
<div
  className="curiosity-usage-graph__totals"
  data-test="graphMetricTotals-hello-world"
>
  <div>
    <div
      className="curiosity-usage-graph__totals-column"
    >
      <Card
        className="curiosity-usage-graph__totals-column-card "
        data-test="graphMetricTotalsCard-0"
        isPlain={true}
      >
        <CardHeader>
          <CardTitle>
            <Title
              headingLevel="h2"
              size="md"
            >
              <Loader
                skeletonProps={
                  {
                    "size": "lg",
                  }
                }
                tableProps={{}}
                variant="skeleton"
              />
            </Title>
          </CardTitle>
        </CardHeader>
        <CardBody>
          <div>
            <Loader
              skeletonProps={
                {
                  "height": "60px",
                  "size": "lg",
                }
              }
              tableProps={{}}
              variant="skeleton"
            />
          </div>
        </CardBody>
        <CardFooter>
          <div>
            <Loader
              skeletonProps={
                {
                  "size": "lg",
                }
              }
              tableProps={{}}
              variant="skeleton"
            />
          </div>
        </CardFooter>
      </Card>
    </div>
  </div>
  <div>
    <div
      className="curiosity-usage-graph__totals-graph-column"
    />
  </div>
</div>
`;

exports[`GraphCardMetricTotals Component should render a basic component: basic 1`] = `""`;
